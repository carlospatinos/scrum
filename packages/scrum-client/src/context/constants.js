const ACTIONS_FETCH = 'FETCH';
const ACTIONS_POST = 'POST';
const ACTIONS_FAILURE = 'FAILURE';
const ACTIONS_SUCCESS = 'SUCCESS';
const LIBRARY_ROOT = 'SCRUM';
const ROUTE_CHANGED = 'ROUTE_CHANGED';

export {
  ACTIONS_FETCH,
  ACTIONS_POST,
  ACTIONS_FAILURE,
  ACTIONS_SUCCESS,
  LIBRARY_ROOT,
  ROUTE_CHANGED,
};
