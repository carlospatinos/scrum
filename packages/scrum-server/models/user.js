const mongoose = require('mongoose');
const UserSchema = new mongoose.Schema({
    name: {
        type: String,
        required: true
    },
    email: {
        type: String,
        required: true
    },
    password: {
        type: String,
        required: true
    },
    userType: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'UserType',
        required: true
    },
    team: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Team',
        required: false
    },
    date: {
        type: Date,
        default: Date.now
    }
});
const User = mongoose.model('User', UserSchema);

module.exports = User;