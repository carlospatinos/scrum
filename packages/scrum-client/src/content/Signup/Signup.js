import React, { useState } from "react";
import { Alert, Button, FormGroup, FormControl, FormLabel } from "react-bootstrap";
import "./Signup.css";
import { useHistory } from 'react-router-dom';
import { API_BASE_URL } from '../../constants/apiConstants';

// import TagManager from 'react-gtm-module'
// if (process.env.NODE_ENV === "development" && !!process.env.REACT_APP_GTM_ID) {
//   document.title = "signup";
//   console.log(document.title);
//   const tagManagerArgs = {
//     dataLayer: {
//       page: "signup", //Specific to each page
//       pagePath: window.location.pathname + window.location.search, //"/signup", //Specific to each page
//       title: "signup"
//     },
//     dataLayerName: "PageDataLayer"
//   };
//   TagManager.dataLayer(tagManagerArgs);
// }

export default function Signup() {
  const history = useHistory();

  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [password2, setPassword2] = useState("");

  const [errorMessage, setErrorMessage] = useState('');

  const [apiResponse] = useState('');

  function validateForm() {
    return (
      firstName.length > 0 &&
      lastName.length > 0 &&
      email.length > 0 &&
      password.length > 0 &&
      password === password2
    );
  }

  function handleSubmit(event) {
    event.preventDefault();

    const requestOptions = {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        firstName,
        lastName,
        email,
        password,
        password2,
      }),
    };

    try {
      fetch(API_BASE_URL + "/api/signup", requestOptions)
        .then((response) => response.json())
        .then((data) => {
          if (data.success) {
            history.push("/home");
          } else {
            setApiResponse(data.message);
          }
        });
    } catch (e) {
      console.log("=====> error:" + e);
      setErrorMessage({ error: e });
      // TODO this erro happen if API is not available but business errors like length of password go above. how to handle and display those?
    }
  }

  const isValidForm = validateForm();

  return (
    <div className="Signup">
      <form onSubmit={handleSubmit}>
        <FormGroup controlId="firstName">
          <FormLabel>First Name</FormLabel>
          <FormControl value={firstName} onChange={e => setFirstName(e.target.value)} />
        </FormGroup>
        <FormGroup controlId="lastName">
          <FormLabel>Last Name</FormLabel>
          <FormControl value={lastName} onChange={e => setLastName(e.target.value)} />
        </FormGroup>

        <FormGroup controlId="email">
          <FormLabel>Email</FormLabel>
          <FormControl
            autoFocus
            type="email"
            value={email}
            onChange={e => setEmail(e.target.value)}
          />
        </FormGroup>
        <FormGroup controlId="password">
          <FormLabel>Password</FormLabel>
          <FormControl
            value={password}
            onChange={e => setPassword(e.target.value)}
            type="password"
          />
        </FormGroup>
        <FormGroup controlId="password2">
          <FormLabel>Password2</FormLabel>
          <FormControl
            value={password2}
            onChange={e => setPassword2(e.target.value)}
            type="password"
          />
        </FormGroup>
        {apiResponse && <Alert variant={"danger"}>{apiResponse}</Alert>}
        {errorMessage && <Alert variant={"danger"}>{errorMessage}</Alert>}
        <Button block disabled={!validateForm()} type="submit" variant={isValidForm ? "primary" : "secondary"}>
          Sign up
        </Button>
      </form>
    </div>
  );
}
