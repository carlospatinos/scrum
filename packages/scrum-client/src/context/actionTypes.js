import ContextUtil from './utils';
import { ACTIONS_POST } from './constants';

const LOGIN_ACTIONS = ContextUtil.generateActionTypes({
  featureName: 'LOGIN',
  restVerb: ACTIONS_POST,
});
const LOGOUT_ACTIONS = ContextUtil.generateActionTypes({
  featureName: 'LOGOUT',
});

export { LOGIN_ACTIONS, LOGOUT_ACTIONS };
