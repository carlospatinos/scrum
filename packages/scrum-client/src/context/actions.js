/* eslint-disable no-console */
import { END_POINTS } from 'scrum-common';
import { API_CONSTANTS } from '../constants';

export async function loginUser(dispatch, loginPayload) {
  const requestOptions = {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    // 'Access-Control-Allow-Credentials': true,
    body: JSON.stringify(loginPayload),
  };

  try {
    dispatch({ type: 'REQUEST_LOGIN' });
    const response = await fetch(
      `${API_CONSTANTS.API_BASE_URL}${END_POINTS.AUTH}${END_POINTS.AUTH_LOCAL}`,
      requestOptions
    );
    const data = await response.json();

    if (data.user) {
      dispatch({ type: 'LOGIN_SUCCESS', payload: data });
      localStorage.setItem(API_CONSTANTS.CURRENT_USER, JSON.stringify(data));
      return data;
    }

    dispatch({ type: 'LOGIN_ERROR', error: data.message });
  } catch (error) {
    dispatch({ type: 'LOGIN_ERROR', error });
    console.log(error);
  }
  return undefined;
}

export async function login3ppUser(dispatch) {
  const requestOptions = {
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
    // 'Access-Control-Allow-Credentials': true,
    credentials: 'include',
  };

  try {
    dispatch({ type: 'REQUEST_LOGIN' });
    const response = await fetch(
      `${API_CONSTANTS.API_BASE_URL}${END_POINTS.AUTH}${END_POINTS.LOGIN_SUCCESS}`,
      requestOptions
    );

    const data = await response.json();
    if (data.user) {
      dispatch({ type: 'LOGIN_SUCCESS', payload: data });
      localStorage.setItem(API_CONSTANTS.CURRENT_USER, JSON.stringify(data));
      return data;
    }

    dispatch({ type: 'LOGIN_ERROR', error: data.message });
    console.log(data.message);
  } catch (error) {
    dispatch({ type: 'LOGIN_ERROR', error });
    console.log(error);
  }
  return undefined;
}

export async function logout(dispatch) {
  dispatch({ type: 'LOGOUT' });
  localStorage.removeItem(API_CONSTANTS.CURRENT_USER);
  localStorage.removeItem(API_CONSTANTS.ACCESS_TOKEN_NAME);
  // TODO is this needed? localStorage.clear();
  localStorage.clear();
}
