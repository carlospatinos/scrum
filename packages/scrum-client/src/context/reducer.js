/* eslint-disable no-shadow */
import { CURRENT_USER } from '../constants/apiConstants';

const user = localStorage.getItem(CURRENT_USER)
  ? JSON.parse(localStorage.getItem(CURRENT_USER)).user
  : '';
const token = localStorage.getItem(CURRENT_USER)
  ? JSON.parse(localStorage.getItem(CURRENT_USER)).login_access_token
  : '';

export const initialState = {
  user: '' || user,
  login_access_token: '' || token,
  loading: false,
  errorMessage: null,
};

export const AuthReducer = (initialState, action) => {
  switch (action.type) {
    case 'REQUEST_LOGIN':
      return {
        ...initialState,
        loading: true,
      };
    case 'LOGIN_SUCCESS':
      return {
        ...initialState,
        user: action.payload.user,
        login_access_token: action.payload.login_access_token,
        loading: false,
      };
    case 'LOGOUT':
      return {
        ...initialState,
        user: '',
        login_access_token: '',
      };

    case 'LOGIN_ERROR':
      return {
        ...initialState,
        loading: false,
        errorMessage: action.error,
      };

    default:
      throw new Error(`Unhandled action type: ${action.type}`);
  }
};
