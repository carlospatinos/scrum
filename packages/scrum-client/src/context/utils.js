import { ACTIONS_FETCH, ACTIONS_SUCCESS, ACTIONS_FAILURE, LIBRARY_ROOT } from './constants';
/**
 * Generates an action type by concatenating the library, feature name and key.
 *
 * @param {String} library
 * @param {String} featureName
 * @param {String} key
 */
const generateActionType = (library, featureName, key) => `${library}/${featureName}/${key}`;

/**
 * Generates a partial implementation of `generateActionType` fixing library and featureName.
 *
 * @param {String} library
 * @param {String} featureName
 */
const generateFeatureActionType = (library, featureName) => key =>
  generateActionType(library, featureName, key);

/**
 * Generates the three default action types, baseAction,  successAction and failureAction of a given feature.
 *
 * @param {*} { library = LIBRARY_ROOT, featureName, restVerb = ACTIONS_FETCH }
 * @returns
 */
const generateActionTypes = ({ library = LIBRARY_ROOT, featureName, restVerb = ACTIONS_FETCH }) => {
  const baseActionKey = `${restVerb}_${featureName.toUpperCase()}`;
  const featureActionType = generateFeatureActionType(library, featureName);
  const baseAction = featureActionType(baseActionKey);
  const successAction = featureActionType(`${baseActionKey}_${ACTIONS_SUCCESS}`);
  const failureAction = featureActionType(`${baseActionKey}_${ACTIONS_FAILURE}`);
  return { baseAction, successAction, failureAction };
};
/**
 * Generates a context action, it handles the three action types of a given feature action.
 * - dispatchFxn( type: action.baseAction): Before invoking the `apiFxn`/
 * - dispatchFxn( type: action.successAction): When the response of  `apiFxn` is received.
 * - dispatchFxn( type: action.baseAction):  When an error  of `apiFxn` is received.
 *
 * @param {Function} dispatchFxn - The dispatcher function.
 * @param {Function} apiFxn - The API function to invoke.
 * @param {Object} action - The baseAction, successAction and failureAction of a given feature.
 */
const generateAction = (dispatchFxn, apiFxn, action) => async req => {
  const initialAction = {
    type: action.baseAction,
    payload: {},
  };
  dispatchFxn(initialAction);

  try {
    const response = await apiFxn(req);
    console.log('response.data', response.data);
    dispatchFxn({
      type: action.successAction,
      payload: response.data,
    });
    return response;
  } catch (e) {
    dispatchFxn({
      type: action.failureAction,
      error: e.message,
    });
    return undefined;
  }
};

/**
 * Context utility , it generates actions types and actions.
 */
const ContextUtil = { generateActionType, generateActionTypes, generateAction };
export default ContextUtil;
