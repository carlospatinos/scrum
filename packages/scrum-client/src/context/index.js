import { loginUser, login3ppUser, logout } from './actions';
import { AuthProvider, useAuthDispatch, useAuthState } from './context';

export { AuthProvider, useAuthState, useAuthDispatch, loginUser, login3ppUser, logout };
