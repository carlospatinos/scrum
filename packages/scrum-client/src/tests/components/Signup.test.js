import { render, cleanup } from '../test-utils';
import Signup from '../../content/Signup';
import { useAuthState } from '../../context/context';
import { basicUser } from '../mocks/user';

jest.mock('../../context/context');
afterEach(cleanup);

describe('Signup', () => {
  test('To match snapshot', () => {
    //   mock hook call
    useAuthState.mockReturnValue(basicUser);

    // render the component
    const { asFragment } = render(<Signup />);

    // validates vs snapshot
    expect(asFragment()).toMatchSnapshot();
  });
});
