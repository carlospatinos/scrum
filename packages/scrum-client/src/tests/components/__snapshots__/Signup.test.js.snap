// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup To match snapshot 1`] = `
<DocumentFragment>
  <div
    class="mt-3 shadow-lg p-3 mb-5 bg-white rounded  jumbotron"
  >
    <form
      class=""
    >
      <h1
        class="text-center"
      >
        Sing up
      </h1>
      <div
        class="form-group"
      >
        <input
          class="form-control"
          id="firstName"
          placeholder="First Name"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <input
          class="form-control"
          id="lastName"
          placeholder="Last Name"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <input
          class="form-control"
          id="email"
          placeholder="Email"
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <input
          class="form-control"
          id="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <input
          class="form-control"
          id="confirmPassword"
          placeholder="Confirm password"
          type="password"
          value=""
        />
      </div>
      <button
        class="btn btn-secondary btn-block"
        disabled=""
        type="submit"
      >
        Sing up
      </button>
      <p
        class="forgot-password text-right"
      >
        <a
          href="/login"
        >
          <small>
            Already registered Sign In?
          </small>
        </a>
      </p>
    </form>
  </div>
</DocumentFragment>
`;
